### PR TITLE
56 create sankey diagram starting from loading the list of nlnet repos until we get as many test counts as practical

### PR DESCRIPTION
**Pull request by tnzmnjm**
Hi Julian,

Could you please review the changes I have made and let me know if any improvement is required?

The changes are:
- I have divided the function ` check_and_clean_data(df)` into 2 functions `remove_duplicates(df)` and `remove_null_values(df)` in the `utils/initial_data_preparation.py` script
- There is a work in progress regarding applying these 2 functions on the `original_df`. At the moment I am not applying them. I am importing these functions into the script `src/sankey_diagram_plotly.py` as I need what they return
- The script was saving dataframes for the repos which had more than 10 repos. I have added a section to create a dataframe for all other repo domains which have fewer repos. The dataframe is called `other_domains.csv`. 
- I have added 'repodomain' column to 'original_df' to categorise data by hosting domain platforms
- Implemented new nodes in the Sankey diagram for 'Duplicates', 'No Repo Name'

As per our discussion, more work is required in order to  enable us to proceed with the sankey diagram and adding more nodes (regarding the successful or failing of cloning the repos). This will be addresed in another issue.

Thanks

### CODE REVIEW

File: utils/initial_data_preparation.py [hunk 1]
```diff
@@ -314,3 +336,18 @@ def get_base_repo_url(url):
                 f"Saved {domain} domain DataFrame with "
                 f"{len(domain_df)} entries to {data_folder}"
             )
+
+        # Append DataFrames with less than 10 repositories to the
```
Comment (julianharty) [inline utils/initial_data_preparation.py:340]: This (and line 330 `if len(domain_df) > 10:`) doesn't seem consistent with `domains_with_more_than_four` in the other file in this commit or with the notes in the PR. How about we extract a constant value that's used consistently as the threshold for what gets stored in a named file e.g. github... vs. what gets stored in a composite of other domains that are relatively infrequent?

File: utils/initial_data_preparation.py [hunk 2]
```diff
@@ -314,3 +336,18 @@ def get_base_repo_url(url):
...
+        else:
+            other_domains_df = pd.concat(
+                [other_domains_df, domain_df], ignore_index=True
+            )
+
+    # Save the DataFrame containing domains with less than 10 repositories
+    if not other_domains_df.empty:
+        other_domains_df.to_csv(data_folder / "other_domains.csv", index=False)
+        logger.info(
+            f"Saved DataFrame with domains having less than 10 repositories to: "
```
Comment (julianharty) [inline utils/initial_data_preparation.py:351]: Again, it's probably worth using a consistent constant value for this number (10) throughout our codebase.

**Review by julianharty — COMMENTED** [review]
Some interim quick comments on this draft PR. Thanks for your progress so far on this feature.

File: src/sankey_diagram_plotly.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+# Reading the original data
+original_df = pd.read_csv(working_directory / "data" / "original.csv")
+initial_row_count = original_df.shape[0]
+
+original_df["repodomain"] = original_df["repourl"].str.extract(
+    r"https?://(www\.)?([^/]+)"
+)[1]
+repodomain_counts = original_df["repodomain"].value_counts()
+
+domains_with_more_than_four = repodomain_counts[repodomain_counts > 6].index
```
[the remark this answers is not in the document]
Comment (tnzmnjm) [reply src/sankey_diagram_plotly.py:70]: This is fixed. I was trying different values.

File: utils/initial_data_preparation.py (hunk 1 above)
Comment (tnzmnjm) [reply utils/initial_data_preparation.py:340]: I changed the values/ texts to be 10 or greater than 10 for consistency.

**Pull request merged**